contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -927,6 +928,265 @@ checksum(PG_FUNCTION_ARGS)
+			errmsg("String or binary data would be truncated.")));
+
+	/* resolve the three part name */
+	splited_object_name = split_object_name(input);
+	db_name = splited_object_name[1];
+	schema_name = splited_object_name[2];
+	object_name = splited_object_name[3];
+
+	/* downcase identifier if needed */
+	if (pltsql_case_insensitive_identifiers)
+	{	
+		db_name = downcase_identifier(db_name, strlen(db_name), false, false);
+		schema_name = downcase_identifier(schema_name, strlen(schema_name), false, false);
+		object_name = downcase_identifier(object_name, strlen(object_name), false, false);
+		for (int i = 0; i < 4; i++)
+			pfree(splited_object_name[i]);
+	}
+	else
+		pfree(splited_object_name[0]);
+
+	pfree(input);
+	pfree(splited_object_name);
+
+	/* truncate identifiers if needed */
+	truncate_tsql_identifier(db_name);
+	truncate_tsql_identifier(schema_name);
+	truncate_tsql_identifier(object_name);
+
+	if (!strcmp(db_name, ""))
+		db_name = get_cur_db_name();
+	else if (strcmp(db_name, get_cur_db_name()) && strcmp(db_name, "tempdb"))
+	{
+		/* cross database lookup */
+		int db_id = get_db_id(db_name);
+		if (!DbidIsValid(db_id))
+		{
+			pfree(db_name);
+			pfree(schema_name);
+			pfree(object_name);
+			if (object_type)
+				pfree(object_type);
+			PG_RETURN_NULL();
+		}
+		user_id = GetSessionUserId();
+	}
+
+	/* get physical schema name from logical schema name */
+	if (!strcmp(schema_name, ""))
+	{	
+		/* find the default schema for current user and get physical schema name */
+		const char *user = get_user_for_database(db_name);
+		const char *guest_role_name = get_guest_role_name(db_name);
+		if (!user)
+		{	
+			pfree(db_name);
+			pfree(schema_name);
+			pfree(object_name);
+			if(object_type)
+				pfree(object_type);
+			PG_RETURN_NULL();
+		}
+		else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+		{
+			physical_schema_name = pstrdup(get_dbo_schema_name(db_name));
+		}
+		else
+		{
+			schema_name  = get_authid_user_ext_schema_name((const char *) db_name, user);
+			physical_schema_name = get_physical_schema_name(db_name, schema_name);
+		}
+	}
+	else
+	{
+		physical_schema_name = get_physical_schema_name(db_name, schema_name);
+	}
+
+	/* get schema oid from physical schema name, it will return InvalidOid if user don't have lookup access */
+	schema_oid = get_namespace_oid(physical_schema_name, true);
+
+	/* free unnecessary pointers */
+	pfree(db_name);
+	pfree(schema_name);
+	pfree(physical_schema_name);
+
+	if(!OidIsValid(schema_oid) || pg_namespace_aclcheck(schema_oid, user_id, ACL_USAGE) != ACLCHECK_OK)
+	{
+		pfree(object_name);
+		if (object_type)
+			pfree(object_type);
+		PG_RETURN_NULL();
+	}
+	
+	/* check if looking for temp object */
+	is_temp_object = (object_name[0] == '#'? true: false);
+
+	if (object_type) /* "object_type" is specified in-argument */
+	{	
+		if (is_temp_object)
+		{
+			if (!strcmp(object_type, "S") || !strcmp(object_type, "U") || !strcmp(object_type, "V") ||
```

Comment:
Shouldnt these be case insensitive comparison?
A better solution would be to downcase-truncate object_type as well